src/blockMonitor/blockCache.ts

```diff
@@ -99,110 +92,142 @@ export class BlockCache<TBlock extends IBlockStub> implements ReadOnlyBlockCache
      * Returns true it `block` can be added to the cache, that is if either:
      *   - it is the first block ever seen, or
      *   - its parent is already in the cache, or
-     *   - it is at depth least `this.maxDepth`.
+     *   - it is at exactly `this.maxDepth`.
+     * If not, it can only be added to the pending.
      * @param block
      */
     public canAddBlock(block: Readonly<TBlock>): boolean {
-        return this.isEmpty || this.hasBlock(block.parentHash) || block.number <= this.minHeight;
+        return this.isEmpty || this.hasBlock(block.parentHash) || block.number === this.minHeight;
```

Comment:
I think you can remove canAddBlock - it at least no longer should be public, it's not used externally. To me it's also more readable to call it `isOrphan` and invert the conditions.

```
        const isOrphan = (block: TBlock) => {
            // if the block cache is empty then we dont have knowledge of parents 
            // so by definition a block cannot be an orphan
            if (this.isEmpty) return false;
            // if the parent of this block is in the cache then this block cannot be an orphan
            if (this.hasBlock(block.parentHash)) return false;
            // if this block number equals the min height, then there can be nothing below,
            // just as with isEmpty it cannot be an orphan
            if (block.number === this.minHeight) return false;
            return true;
        };

        if (isOrphan(block)) {
            this.pendingBlocksByHash.set(block.hash, block);
            return false;
        } else {
            this.blocksByHash.set(block.hash, block);

...
```

Reply:
I think it's still useful for testing, as it is still a relatively non-trivial semantics and we don't want it to be changed inadvertently.
In 4fce304 I changed the name to "canAttachBlock" which fits the current attached/detached nomenclature.
(I also removed it from the signature of the ReadOnlyBlockCache, which is what's passed around in the application outside of the BlockProcessor.)
What do you think?

Reply:
Yeah, I think it's fine. Can you include just a bit more comment information like i did about the isEmpty and minHeight cases.

Reply:
Done in a545d55; I added to the comment on top of the function instead of commenting inline, as having both would be redundant.